community/kernel/src/main/java/org/neo4j/helpers/SortedLongArrayUtil.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.helpers;
+
+/**
+ * Specialized methods for operations of sorted long arrays.
+ */
+public class SortedLongArrayUtil
+{
+    private static final long[] EMPTY = new long[]{};
+
+    /**
+     * Compute union of two sorted long arrays.
+     * @param left a sorted array set
+     * @param right another sorted array set
+     * @return the union, which is NOT sorted
+     */
+    public static long[] union( long[] left, long[] right )
+    {
+        if ( left == null || right == null )
+        {
+            return left == null ? right : left;
+        }
+
+        long uniqueCounts = countUnique( left, right );
+        if ( uniqueCounts == 0 || right( uniqueCounts ) == 0 )
+        {
+            return left;
+        }
+        if ( left( uniqueCounts ) == 0 )
+        {
+            return right;
+        }
+
+        long[] union = new long[left.length + right( uniqueCounts )];
+
+        int cursor = 0;
+        int l = 0;
+        int r = 0;
+        while ( l < left.length && r < right.length )
+        {
+            if ( left[l] == right[r] )
+            {
+                union[cursor++] = left[l];
+                l++;
+                r++;
+            }
+            else if ( left[l] < right[r] )
+            {
+                union[cursor++] = left[l];
+                l++;
+            }
+            else
+            {
+                union[cursor++] = right[r];
+                r++;
+            }
+        }
+        while ( l < left.length )
+        {
+            union[cursor++] = left[l];
+            l++;
+        }
+        while ( r < right.length )
+        {
+            union[cursor++] = right[r];
+            r++;
+        }
+
+        assert cursor == union.length;
+        return union;
+    }
+
+    /**
+     * Compute intersect of two sorted long array sets.
+     * @param left a sorted array set
+     * @param right another sorted array set
+     * @return the union, which is NOT sorted
+     */
+    public static long[] intersect( long[] left, long[] right )
```

Comment:
Again, I don't see the point of not returning a sorted array - it would make it more useable if it took two "sets" as input and returned a "set" as output.

This method should of course be moved with its buddy above...